### ISSUE TITLE
onStyleLoad for gathering map ref is not reliable

### ISSUE THREAD

**Issue by cyrilchapon** [alex3165/react-mapbox-gl#601]
There are edge cases where styleLoad is not being fired

For example, adding a WMTS source that returns 404. The `onStyleLoad` will never fire.

Among being weird for mapboxgl, relying on that event for getting map instance is not sufficient. I personally need the map sooner. Thinking of it, actually, waiting for the map being, kinda, "ready" to expose the instance is restrictive and not very consistent for a wrapper.

Maybe something like a `ref=` callback in react would be great.

**Comment by renchap** [comment]
I agree, a `mapRef` Prop on the component would be very useful for advanced uses.

@alex3165 would you be open to a PR adding this feature?

**Comment by mklopets** [comment]
@renchap I'm not Alex but I'd say a PR would be welcome indeed 🙂

**Comment by MomoRazor** [comment]
Hi there, just stumbled on this issue myself, (need map instance, but onStyleLoad isn't firing for me), and I was wondering since this thread is rather old, if there have been any alternatives for this, since no PR seems to have happened here. Thanks!

**Comment by yhrchan** [comment]
Hi,

I am using https://github.com/amaurymartiny/react-mapbox-gl-draw in conjunction with react-mapbox-gl in which I need to access the map instance from within a defined "onDrawCreate" event. However, I've tried different ways and have been exhausted all the options. If anyone knows of a solution I would appreciate any help or suggestions.

Thanks!
